internal/builder/resourceset.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2024 Stefan Prodan.
+// SPDX-License-Identifier: AGPL-3.0
+
+package builder
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/template"
+
+	ssautil "github.com/fluxcd/pkg/ssa/utils"
+	sprig "github.com/go-task/slim-sprig/v3"
+	"github.com/gosimple/slug"
+	apix "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/yaml"
+
+	fluxcdv1 "github.com/controlplaneio-fluxcd/flux-operator/api/v1"
+)
+
+// BuildResourceSet builds a list of Kubernetes resources
+// from a list of JSON templates using the provided inputs.
+func BuildResourceSet(yamlTemplate string, templates []*apix.JSON, jsonInputs []fluxcdv1.ResourceSetInput) ([]*unstructured.Unstructured, error) {
+	inputs := make([]map[string]any, 0, len(jsonInputs))
+	for i, ji := range jsonInputs {
+		inp := make(map[string]any, len(ji))
+		for k, v := range ji {
+			var data any
+			if err := json.Unmarshal(v.Raw, &data); err != nil {
+				return nil, fmt.Errorf("failed to unmarshal inputs[%d]: %w", i, err)
+			}
+			inp[k] = data
+		}
+		inputs = append(inputs, inp)
+	}
+
+	var objects []*unstructured.Unstructured
+
+	// build resources from JSON templates
+	for i, tmpl := range templates {
+		if len(inputs) == 0 {
+			object, err := BuildResource(tmpl, nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to build resource: %w", err)
+			}
+
+			objects = append(objects, object)
+			continue
+		}
+
+		for _, input := range inputs {
+			object, err := BuildResource(tmpl, input)
+			if err != nil {
+				return nil, fmt.Errorf("failed to build resources[%d]: %w", i, err)
+			}
+
+			// exclude object based on annotations
+			if val := object.GetAnnotations()[fluxcdv1.ReconcileAnnotation]; val == fluxcdv1.DisabledValue {
+				continue
+			}
+
+			// deduplicate objects
+			if !containsObject(objects, object) {
+				objects = append(objects, object)
+			}
+		}
+	}
+
+	// build resources from multi-doc YAML template
+	if yamlTemplate != "" {
+		var objectsFromTemplate []*unstructured.Unstructured
+		if len(inputs) == 0 {
+			objs, err := BuildResourcesFromYAML(yamlTemplate, nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to build resource: %w", err)
+			}
+
+			objectsFromTemplate = append(objectsFromTemplate, objs...)
+		}
+		for _, input := range inputs {
+			objs, err := BuildResourcesFromYAML(yamlTemplate, input)
```

Comment:
Also I just realized that we can build and parse the template just once in this case and execute it for each `input`.

Reply:
Actually this also true for `.spec.resources`.

Reply:
Hmm not really, inputs can differ in structure.

Reply:
Oh yeah, totally, my bad!